File: 322-coin-change/322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        //tabulation dp
        int n = coins.length;
        int[][] dp = new int[n + 1][amount + 1];
        
        for(int i=0;i<=n;i++)
            Arrays.fill(dp[i], -1);
        
        int minCoins = tabulation(coins, amount, coins.length, dp);
        return minCoins == Integer.MAX_VALUE - 1 ? -1 : minCoins;
    }
    
    int tabulation(int[] coins, int amount, int n, int[][] dp) {
        //initilizing the first row with inf.
        for(int i=1;i<=amount;i++)
            dp[0][i] = Integer.MAX_VALUE - 1;
        
        //initilizing the first col with 0.
        for(int i=0;i<=n;i++)
            dp[i][0] = 0;
        
        for(int i=1;i<dp.length;i++) {
            for(int j=1;j<dp[0].length;j++) {
                if(coins[i - 1] <= j)
                    dp[i][j] = Math.min(dp[i - 1][j], 1 + dp[i][j - coins[i - 1]]);
                else
                    dp[i][j] = dp[i - 1][j];
            }
        }
        return dp[n][amount];
    }
}

File: 322-coin-change/322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp(amount + 1, vector<int>(n + 1, -1));
        int minCoins = helper(coins.size(), coins, amount, dp);
        
        return minCoins == INT_MAX - 1 ? -1 : minCoins;
    }
    
    int helper(int index, vector<int>& coins, int amount, vector<vector<int>>& dp) {
        if(index == 0 || amount == 0) 
            return (amount == 0) ? 0 : INT_MAX - 1;
        
        if(dp[amount][index] != -1)
            return dp[amount][index];
        
        if(coins[index - 1] > amount) 
            return dp[amount][index] = helper(index - 1, coins, amount, dp);
        else 
            return dp[amount][index] = min(1 + helper(index, coins, amount - coins[index - 1], dp), helper(index - 1, coins, amount, dp));
    }
};